### ISSUE TITLE
[DÉFI 21] LE NIMBUS 3000 - Benchmark d'optimizers

### ISSUE THREAD

**Issue by juju149** [juju149/workshop-poudlard-epsi#45]
## 🎯 Objectif

Réaliser un **benchmark rigoureux** des optimizers suivants sur un réseau existant (ex : réseau du défi *Is iT you hARrY* ou tout autre modèle public) :

* Adadelta
* Adagrad
* RMSProp
* Adam
* SGD (avec et sans momentum / Nesterov)
* Adan
* AdamW *(et variante AdamV2 si disponible dans le framework)*

> *But : produire un **rapport au format papier de recherche** qui démontre quels optimizers sont les plus adaptés au réseau choisi, avec une méthodologie reproductible et des scripts d’exécution.*

---

## 📦 Livrables attendus

* 📘 **Rapport académique (PDF)** : contexte, méthodo, protocole expérimental, résultats, analyses, limites, travaux futurs.
* 🧪 **Scripts de reproduction** : code pour lancer les entraînements, agréger les résultats, générer les tableaux/figures.

---

## 👥 Copilots recommandés

* 🧠 **AI Copilot (lead)**
* 📊 **Data Copilot**

---

## ⏱️ Estimation

* **Story Points** : 5
* **Deadline** : 15/10/2025

---

## 🔭 Périmètre & hypothèses

* **Réseau** : reprendre un modèle existant (CNN/Transformer/MLP) déjà fonctionnel.
* **Données** : utiliser un dataset public ou celui du défi *(Is iT you hARrY)*.
* **Environnement** : PyTorch ou TensorFlow/Keras, CUDA si disponible.
* **Objectif mesuré** : vitesse de convergence, score final (ex. F1/AUROC/Accuracy), stabilité (variance inter-runs), coût compute.

---

### Modèle

* Décrire l’architecture (nb couches, params, activation, normalisation, dropout).
* Geler/dégeler certaines couches si transfert learning.

### Prétraitement & split

* Split fixe **train/val/test** (ex. 70/15/15) + **k-fold** en option.
* Normalisation/standardisation documentée et identique pour tous les runs.

### Métriques

* **Classification** : Accuracy, F1 (macro), AUROC, LogLoss, Calibration (ECE).
* **Régression** : MSE, MAE, R².
* **Ressources** : temps/epoch, temps total, conso VRAM, nombre d’updates.

### Protocole d’entraînement (constant entre optimizers)

* **Époques** : p.ex. 50–100 (avec early stopping patience 10).
* **Batch size** : fixe (ex. 64).
* **Initialisation** : même seed, même init (xavier/kaiming).
* **Scheduler** : Cosine Annealing / StepLR / ReduceLROnPlateau (identique pour tous ou off).
* **Regularisation** : weight decay (documenter), dropout constant, gradient clipping (ex. 1.0).
* **AMP** : mixed precision on/off (constant).
* **Re-runs** : **n=5 seeds** par optimizer pour variance.

### Grille d’hyperparamètres par optimizer

> Chaque optimizer aura **une petite grille** (2–4 configs) + **sélection par validation**.

* **SGD** : lr ∈ {0.1, 0.01}, momentum ∈ {0, 0.9}, nesterov ∈ {true,false}
* **Adam** : lr ∈ {1e-3, 3e-4}, betas ∈ {(0.9,0.999),(0.9,0.95)}
* **AdamW** : lr ∈ {1e-3, 3e-4}, weight_decay ∈ {0.01, 0.05}
* **RMSProp** : lr ∈ {1e-3, 3e-4}, alpha ∈ {0.9, 0.95}, centered ∈ {false,true}
* **Adagrad** : lr ∈ {1e-2,1e-3}, initial_accumulator_value ∈ {0, 0.1}
* **Adadelta** : lr ∈ {1.0, 0.5}, rho ∈ {0.9, 0.95}
* **Adan** : lr ∈ {1e-3, 3e-4}, weight_decay ∈ {0.01, 0.02}

> *Remarque : si votre framework expose **AdamV2**, ajouter une variante avec defaults améliorés (bias correction / eps).*

---

## 🧰 Organisation du repo & scripts

```
optim-benchmark/
├─ data/
│  ├─ raw/              # jeu de données brut
│  └─ processed/        # tensors/records prêts
├─ src/
│  ├─ models/
│  │  └─ net.py         # modèle sélectionné
│  ├─ datasets/
│  │  └─ loader.py      # dataloaders + transforms
│  ├─ train.py          # boucle d'entraînement générique
│  ├─ eval.py           # éval test + agrégation
│  ├─ utils.py          # seed, metrics, logging
│  └─ optim_space.py    # définitions des grilles par optimizer
├─ runs/
│  └─ logs/             # CSV/JSONL de mesures par epoch
├─ reports/
│  ├─ figures/          # courbes loss/metric, tables
│  └─ paper.md          # version Markdown avant PDF
├─ scripts/
│  ├─ run_grid.sh       # lance toutes les grilles
│  ├─ summarize.py      # compile résultats + stats
│  └─ plot_curves.py    # génère figures publiables
├─ requirements.txt
└─ README.md
```

### Exemples de commandes

```bash
# 1) Entraîner pour un optimizer donné avec sa grille
python src/train.py --optimizer adamw --grid small --epochs 80 --seed 42

# 2) Lancer toutes les expériences (tous optimizers + seeds)
bash scripts/run_grid.sh

# 3) Agréger et calculer moyennes/écarts
python scripts/summarize.py --runs runs/logs --out reports/summary.csv

# 4) Générer les figures & tableaux du papier
python scripts/plot_curves.py --in reports/summary.csv --out reports/figures
```

---

## 📈 Rapport académique – structure attendue

1. **Résumé** (10 lignes)
2. **Introduction** : contexte, problématique, contributions
3. **Travaux connexes** : panorama rapide des optimizers
4. **Méthodologie** : données, modèle, protocole, grilles, métriques
5. **Résultats** : tableaux (moyenne ± écart-type), courbes de convergence, statistiques (t-test/ANOVA si pertinent)
6. **Analyse** : vitesse vs performance, stabilité, sensibilité aux hyperparams, coût compute
7. **Limites** : taille dataset, compute, généralisation à d’autres tâches
8. **Conclusion & travaux futurs** : recommandations d’optimizers selon cas d’usage
9. **Annexes** : configs exactes, seeds, versions, commande W&B/MLflow

---

## 🔁 Reproductibilité & suivi

* **Seeds fixés** (liste documentée), versions (CUDA, cuDNN, PyTorch/TF).
* **Suivi expérimental** : CSV/JSONL + **Weights & Biases** ou **MLflow** en option.
* **Baseline gelée** : une config de référence commune à tous les runs.
* **CI rapide** : test de fumée (1 epoch, tiny split) pour valider que chaque optimizer s’exécute.

---

## ✅ Critères d’acceptation

* [ ] Toutes les expériences exécutées pour **tous** les optimizers listés.
* [ ] **≥ 5 seeds** par optimizer, mêmes splits, même protocole.
* [ ] Tableaux récapitulatifs + figures (loss/metric vs epoch).
* [ ] Rapport PDF conforme à la structure ci-dessus.
* [ ] Scripts **one-command** pour reproduire les résultats.
* [ ] README clair (prérequis, installation, étapes).

---

## 🧭 Roadmap rapide

* J0–J1 : choix modèle/dataset, setup repo & scripts
* J2–J3 : exécutions grille + logging
* J4 : agrégation résultats, figures, rédaction du papier
* J5 : relectures, packaging final (PDF + scripts)

---

## 📝 Notes

* Si « AdamVV » a été mentionné par erreur, nous couvrons **AdamW** et la **variante AdamV2** si le framework le propose.
* Ajouter en option **Adamax** / **AdaBelief** si temps disponible (section « travaux futurs »).
